### ISSUE TITLE
RSpec Testing with Sidekiq::Testing.disable! is not working for jobs scheduled in the future | How to properly test enqueueing and dequeueing of jobs in test environment

### ISSUE THREAD

**Issue by alexvbush** [mperham/sidekiq#2975]
Hi, I was working on a feature and wanted to unit-test sidekiq but it was not that straightforward. 
I followed this wiki https://github.com/mperham/sidekiq/wiki/Testing and tried to use `Sidekiq::Testing.fake!` at first which will enqueue new jobs onto `jobs` array on worker's class. But that wasn't reflecting the reality of what actually is going on in my codebase. Enqueuing works and can verify it by asserting those worker class arrays but dequeueing is using `Sidekiq::Queue.new` under the hood which won't affect worker's `jobs` array at all in my test environment i.e. I add a job using `perform_at` and that worker's `jobs` array has a new `jid` in it but when I dequeue that jobs using:
```
Sidekiq::Queue.new.each do |job|
     job.delete if reservation.notification_jobs.include?(job.jid)
end
```
it won't remove that `jid` from the array therefor not correctly reflecting what should actually happen.

After that I tried to user `Sidekiq::Testing.disable!` to more closely mimic queues in my test environment (as far as I understand this should be as close as it gets since we're using redis under the hood, just not performing actual jobs upon enqueueing them). But it turns out that jobs scheduled in the future using `perform_at` won't get enqueued. I get a `jid` as return value of `perform_at` method call but when I examine `expect(Sidekiq::Queue.new.size).to eq 1` to returns zero. 
Luckily the same `perform_at` code works as expected in production environment for jobs scheduled in the future and I was able to go move on but currently I'm not fully testing and asserting jobs in my specs due to the issue described here.

Am I using it right or there's another way to do it in test environment? If not than I believe there's a bug with `Sidekiq::Testing.disable!`.

**Comment by mperham** [comment]
I can't follow this issue at all.  Dumb it down please.

**Comment by alexvbush** [comment]
@mperham enqueueing and dequeueing with `Sidekiq::Testing.disable!` in test environment isn't returning right results for `Sidekiq::Queue.new.size`

**Comment by alexvbush** [comment]
@mperham when you schedule jobs in the future with `perform_at `

**Comment by mperham** [comment]
Yes, the API does not have a testing mode.  It always shows you what is really in Redis.

**Comment by alexvbush** [comment]
interesting enough it returns correct results when `perform_at ` schedules something in the past

**Comment by mperham** [comment]
Yes, because of this: https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/worker.rb#L72

**Comment by jagthedrummer** [comment]
@mperham Just to clarify does "the API does not have a testing mode" translate to "it's impossible to test that jobs are queued in the future"?

**Comment by mperham** [comment]
Nope, you can go iterating through jobs in Redis if you want to.

**Comment by jagthedrummer** [comment]
Cool, good news. Could you possibly point me to an example of how I'd go about that?

**Comment by mperham** [comment]
https://github.com/mperham/sidekiq/wiki/API#scheduled

**Comment by jagthedrummer** [comment]
Derp! I thought I had looked everywhere, but I obviously missed that. Thank you!
